### ISSUE TITLE
Getting gravitational waves with s2fft

### ISSUE THREAD

**Issue by robynlm** [astro-informatics/s2fft#324]
Hello, 

I would like to be able to calculate gravitational waves waveforms using s2fft. In [this gist](https://gist.github.com/robynlm/884ee9544f383fc37b5e407efbc7b3dd) you can see how I went about trying this.

I get Psi_4 from simulations, interpolate this on a sphere with the relevant sampling then ask s2fft to get the -2 spin weighted harmonics with `forward`. Repeat this process over time to get the waveform in the different modes. Then I compare to the results from [Multipole](https://einsteintoolkit.org/thornguide/EinsteinAnalysis/Multipole/documentation.html) from Einstein Toolkit. These use different interpolation, sampling and integration so I do not expect them to match exactly. 

So I see similarities in the 22, and 44, and the real part of the 20 mode, some are just numerical noise, others are not matching up. 
So questions:
Am I using s2fft in the way that it was designed to be used? Or am I not calling it right? 
Shouldn’t 2 -2 and 4 -4 be similar to 22 and 44?
Am I calling the l and m moments correctly? 
Am I passing the real/complex parts to s2fft correctly? 
Is there something else I should be doing because it’s a spin -2 field?
Or do you recon something is off in how I reconstruct Psi_4 or interpolate?

Thank you!
Robyn

**Comment by matt-graham** [comment]
Hi @robynlm. 

I'm not sure but I suspect your interpolation step might be causing quite a large loss of information from the `Psi_4` field data. For your chosen value of `lmax = 5` which you are using to set the harmonic bandlimit parameter `L` in the lines calculating the colatitude $\theta$ and longitude $\phi$ points to interpolate at

```Python
sampling = "mw"
lmax = 5
theta = s2fft.sampling.s2_samples.thetas(L=lmax, sampling=sampling)
phi = s2fft.sampling.s2_samples.phis_equiang(L=lmax, sampling=sampling)
```

will result in a grid of 5 colatitude $\theta$ values and 9 longitude $\phi$ values. Looking at the plots of the simulated `Psi_4` fields in the notebook I suspect the original grid the data is sampled on is much denser than a $5 \times 9$ grid? Interpolating to a much coarser grid and then computing harmonic coefficients using `s2fft.forward` will likely give results quite different from a forward transform computed on the pre-interpolation sampled data (assuming this is what Multipole is doing).

Note that the `L` value you use to construct the sampling scheme / grid you interpolate on does not need to match to the `L` you use in computing the forward transform, so you could interpolate to a much finer resolution grid and still only compute the harmonic coefficients up to `L = 5`. 

Also just to note that what we denote as `L` in S2FFT is the non-inclusive upper bound on the order of spherical harmonics to include in expansion - that is we assume a bandlimited signal $f : \mathbb{S} \to \mathbb{C}$ can be represented

$$f(\theta, \phi) = \sum_{\ell=0}^{L-1} \sum_{m=-\ell}^\ell \hat{f}\_{\ell,m} Y\_{\ell,m}(\theta, \phi)$$

which differs from what is sometimes denoted $\ell_{\text{max}}$ which instead corresponds to $\ell_{\text{max}} = L - 1$.

As another comment - you are currently computing a forward transform for a sequence of `psi4_sphere` arrays in a for loop and then manually joining together the coefficients for each `l, m` pair into a list within a dictionary keyed by the `l, m` tuple. You could instead do this as a batch of forward transforms all at once with something like

```Python
import numpy as np
import jax
import s2fft
jax.config.update("jax_enable_x64", True)

...

psi4_sphere_fields = []
# Calc harmonics for each iteration
for it_index in range(len(data['it'])):

    # Reconstruct full box
    coord, Psi4r = fd.reconstruct(data["Weyl_Psi4r"][it_index])
    coord, Psi4i = fd.reconstruct(data["Weyl_Psi4i"][it_index])

    # Interpolation functions
    interp_real = scipy.interpolate.RegularGridInterpolator(coord, Psi4r)
    interp_imag = scipy.interpolate.RegularGridInterpolator(coord, Psi4i)
    psi4_sphere = (interp_real(points_sphere) + 1j * interp_imag(points_sphere))
    psi4_sphere_fields.append(psi4_sphere.reshape(theta_sphere.shape))

psi4_sphere_fields = np.stack(psi4_sphere_fields)
psilm = jax.vmap(
    lambda psi4_sphere: s2fft.forward(psi4_sphere, L=lmax, spin=-2, sampling=sampling, method="jax")
)(psi4_sphere_fields)
```

Note that here `psilm` will be an array with first dimension equal in size to `len(data['it'])` and `psilm[it_index, :, :]` will be the harmonic coefficients for `it_index` arranged using S2FFT's layout convention (row index corresponding to order `l`, column index corresponding to `L - 1 + m`).

**Comment by robynlm** [comment]
I figured it out, it was a me issue in how I was reconstructing the data for a full sphere to interpolate from the symmetric/rotation boundary conditions. Thanks for your help though!
